### PR TITLE
LONDON | May-2025 | Sisay Mehari | Module-Structuring-and-Testing-Data | Sprint-3

### PR DESCRIPTION
**Pull request by sisaymehari**
Learners, PR Template

Self checklist

- [X] I have committed my files one by one, on purpose, and for a reason
- [X] I have titled my PR with REGION | COHORT_NAME | FIRST_NAME LAST_NAME | PROJ_NAME
- [X] I have tested my changes
- [X] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [X] My changes meet the [requirements](./README.md) of this task

Changelist
This PR includes all the exercises I worked on for Sprint-3. I have completed tasks across the following sections:

Implemented functions and wrote tests for find.js, password-validator.js, and related test files.

Completed exercises involving string manipulation, validation logic, and Jest test writing.

Debugged and improved code readability with comments and descriptive naming.

Questions
None at the moment.

### CODE REVIEW

**Review by rohan-corbyn — CHANGES_REQUESTED** [review]
Well done - Great work so far! You've shown a clear understanding of all the concepts introduced in this chapter and you've provided well written and structured solutions. Just a few suggestions here and there.

File: Sprint-3/1-key-implement/1-get-angle-type.js [hunk 1]
```diff
@@ -1,56 +1,88 @@
...
 // The first test and case is written for you. The next case has a test, but no code.
-// Execute this script in your terminal
-// node 1-get-angle-type.js
+// Execute this script in your terminal: `node 1-get-angle-type.js`
 // The assertion error will tell you what the expected output is
 // Write the code to pass the test
 // Then, write the next test! :) Go through this process until all the cases are implemented
 
+// Function to return the type of angle based on its degree
 function getAngleType(angle) {
```
[the remark this answers is not in the document]
Comment (Sisu860) [reply Sprint-3/1-key-implement/1-get-angle-type.js:10]: -What bug would be introduced?
=>multiple result might get logged, because without the return, the function doesn’t exit after one condition matches.
 it breaks the logic, because now conditions aren't mutually exclusive in execution.
-Would it be just as efficient?
=>no . we're evaluating more conditions than needed now, because the function doesn’t exit early anymore. So it’s both logically incorrect and less efficient.
-Could a different type of conditional statement be used to ensure that a change like this doesn't impact the logic or efficiency of the function?
=> Using else if instead of separate ifs would avoid this, since only one branch can run in that structure, it prevents the bug and makes the code more adaptable  to this kind of change. It also makes the conditions mutual exclusivity.

File: Sprint-3/3-mandatory-practice/implement/count.test.js [hunk 2]
```diff
@@ -1,24 +1,60 @@
...
+});
+
+// Scenario: Empty String
+test("should return 0 for an empty string input", () => {
+  expect(countChar("", "a")).toEqual(0);
+  expect(countChar("", " ")).toEqual(0);
+  expect(countChar("", "z")).toEqual(0);
+});
+
+// Scenario: Empty Search Character (Edge Case / Unusual Input)
```
[the remark this answers is not in the document]
Comment (Sisu860) [reply Sprint-3/3-mandatory-practice/implement/count.test.js:43]: -The Empty Search Character  scenario isn’t mentioned in the instructions. it’s more of an edge case I added to see how the function behaves with unusual input.
-When splitting a string by an empty string, it basically splits between every character, so "hello".split('') results in an array with length 6 (["h", "e", "l", "l", "o", ""]). Subtracting 1 gives 5, which matches the number of characters in the string.
If it’s confusing or not relevant, I’m happy to remove it.

File: Sprint-3/3-mandatory-practice/implement/repeat.js [hunk 3]
```diff
@@ -1,5 +1,26 @@
...
+function repeat(str, count) {
+    // Case: Handle Negative Count
+    // If count is less than 0, it's not a valid number of times to repeat.
+    if (count < 0) {
+        throw new Error("Count must be non-negative");
+    }
+
+    // Case: Handle Count of 0
+    // If count is 0, the string should be repeated zero times, resulting in an empty string.
+    if (count === 0) {
```
[the remark this answers is not in the document]
Comment (Sisu860) [reply Sprint-3/3-mandatory-practice/implement/repeat.js]: It's not technically necessary,  I had included it for clarity.
Nothing happen function would still behave correctly without it. the loop would naturally skip and return an empty string.

**Comment by Sisu860** [comment]
I’ve completed the arrangement as requested. Could you please review it and, if everything looks good, mark it as complete so I can submit?

File: Sprint-3/3-mandatory-practice/implement/count.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohan-corbyn) [reply Sprint-3/3-mandatory-practice/implement/count.test.js:43]: No problem at all. 
You could consider null, undefined, 0 or throwing an exception to handle such edge-cases. 

If someone uses the function **wrong** such as:
- Not passing it a required variable.
- Passing the wrong type of variable.
- Passing it a variable outside of conceptual constraints such as > 360° angle.
..they should be notified and shouldn't have to read the code to understand what they did wrong.

**Review by rohan-corbyn — APPROVED** [review]

**Comment by rohan-corbyn** [comment]
There are a a couple changes you might consider which we have discussed already but you have understood the exercises and topics very well and have taken time to work on my suggestions. Good work! 👍

**Comment by Sisu860** [comment]
Thanks so much! I really appreciate your feedback and guidance and I’m glad to hear I’m on the right track. I’ll keep applying your suggestions!
